pkg/binlog/reader/tcp.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reader
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/pingcap/errors"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/utils"
+)
+
+// TCPReader is a binlog event reader which read binlog events from a TCP stream.
+type TCPReader struct {
+	stageMu sync.Mutex
+	stage   readerStage
+
+	syncerCfg replication.BinlogSyncerConfig
+	syncer    *replication.BinlogSyncer
+	streamer  *replication.BinlogStreamer
+}
+
+// TCPReaderStatus represents the status of a TCPReader.
+type TCPReaderStatus struct {
+	Stage      string `json:"stage"`
+	Connection uint32 `json:"connection"`
+}
+
+// String implements Stringer.String.
+func (s *TCPReaderStatus) String() string {
+	data, err := json.Marshal(s)
+	if err != nil {
+		log.Errorf("[TCPReaderStatus] marshal status to json error %v", err)
+	}
+	return string(data)
+}
+
+// NewTCPReader creates a TCPReader instance.
+func NewTCPReader(syncerCfg replication.BinlogSyncerConfig) Reader {
+	return &TCPReader{
+		syncerCfg: syncerCfg,
+		syncer:    replication.NewBinlogSyncer(syncerCfg),
+	}
+}
+
+// StartSyncByPos implements Reader.StartSyncByPos.
+func (r *TCPReader) StartSyncByPos(pos gmysql.Position) error {
+	r.stageMu.Lock()
+	defer r.stageMu.Unlock()
+
+	if r.stage != stageNew {
+		return errors.NotValidf("stage %d, expect %d", r.stage, stageNew)
```

Comment:
is log `stage %d, expect %d is not valid`?

Reply:
changed to `stage %s, expect %s` only.